### PR TITLE
SQL-1093: audit to_x for i32, i64, u32, and u64

### PR DESCRIPTION
**Pull request by terakilobyte**
I've audited the tests, added more, and refined some error messages.

Per the spec (as I understand it), integer truncation is a "loss of the whole" and is an error rather than an Ok with a result and warning.

Interestingly, I'm getting intermittent test failures for test panic. Would appreciate a point out to where it might be.

```sh
cargo test --package mongo-odbc-driver --lib

...
failures:

---- api::panic_safe_exec_tests::unit::test_panic stdout ----
thread 'api::panic_safe_exec_tests::unit::test_panic' panicked at 'assertion failed: `(left == right)`
  left: `"Panic(\"panic test\\nOk(\\\"in file '"`,
 right: `"Panic(\"panic test\\nErr(RecvError)"`', odbc/src/api/panic_safe_exec_tests.rs:61:13
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace


failures:
    api::panic_safe_exec_tests::unit::test_panic
```

### CODE REVIEW

File: odbc/src/api/data.rs [hunk 1]
```diff
@@ -140,29 +142,89 @@ impl IntoCData for Bson {
...
+                Err(ODBCError::IntegralTruncation(x.to_string()))
+            }
+            Bson::Int64(x) if *x > u32::MAX as i64 || *x < 0i64 => {
+                Err(ODBCError::IntegralTruncation(x.to_string()))
+            }
+            Bson::Decimal128(x) if self.to_f64()? > u32::MAX as f64 || self.to_f64()? < 0f64 => {
+                Err(ODBCError::IntegralTruncation(x.to_string()))
+            }
+            Bson::Int32(x) if *x < 0i32 => Err(ODBCError::IntegralTruncation(x.to_string())),
+            _ => self.to_u64().map(|(u, w)| (u as u32, w)),
```
Comment (pmeredit) [inline odbc/src/api/data.rs:226]: Maybe I'm missing something here, but we seem to be calling `to_u64` twice. The first time we just throw the result away.

File: odbc/src/api/data.rs [hunk 2]
```diff
@@ -140,29 +142,89 @@ impl IntoCData for Bson {
         }
     }
 
-    fn to_u64(&self) -> Result<(u64, Option<ODBCError>)> {
-        self.to_i64().map(|(i, w)| (i as u64, w))
-    }
-
     fn to_i32(&self) -> Result<(i32, Option<ODBCError>)> {
-        let (out_64, warning) = self.to_i64().map_err(|e| match e {
+        self.to_i64().map_err(|e| match e {
```
Comment (pmeredit) [inline odbc/src/api/data.rs:146]: We seem to be calling to_i64 twice, throwing the result away the first time.

**Review by pmeredit — CHANGES_REQUESTED** [review]
Just a small change, unless I'm wrong?

File: odbc/src/api/data.rs [hunk 3]
```diff
@@ -140,29 +142,89 @@ impl IntoCData for Bson {
...
-        Ok((
-            out,
-            warning.or(if out as i64 != out_64 {
-                Some(ODBCError::IntegralTruncation(self.to_string()))
-            } else {
-                None
-            }),
-        ))
+        match self {
+            Bson::Double(x) if *x > i32::MAX as f64 => {
```
Comment (jchemburkar) [inline odbc/src/api/data.rs:152]: is the inverse of this possible (e.g. `*x < i32::MIN`)? What would be the expected behavior in that case?

File: odbc/src/api/data.rs [hunk 4]
```diff
@@ -1141,6 +1203,183 @@ mod unit {
         );
     }
 
+    mod conversion {
```
Comment (jchemburkar) [inline odbc/src/api/data.rs:1206]: I like all the tests added below -- the naming convention was very helpful too!

File: odbc/src/api/data.rs (hunk 1 above)
Comment (terakilobyte) [reply odbc/src/api/data.rs:226]: I was doing this to check for unsupported conversions. Thanks for calling this out. I'm refactoring it to `map_or_else` which will provide a much cleaner implementation!

**Review by jchemburkar — COMMENTED** [review]
Looks good, and I thought the difference between Fractional/Integral truncation made sense when reading through. I left one question about an edge case, but really liked the added tests and appreciated the verbosity in the naming!

File: odbc/src/api/data.rs (hunk 3 above)
Comment (terakilobyte) [reply odbc/src/api/data.rs:147]: That's a great point @jchemburkar. Good catch!

**Review by jchemburkar — APPROVED** [review]
lgtm!

File: odbc/src/api/data.rs [hunk 5]
```diff
@@ -129,28 +130,34 @@ impl IntoCData for Bson {
...
+            Bson::Int64(x) if *x > i32::MAX as i64 || *x < i32::MIN as i64 => {
                 Err(ODBCError::IntegralTruncation(x.to_string()))
             }
-            Bson::Decimal128(x) if self.to_f64()? > i32::MAX as f64 => {
+            Bson::Decimal128(x)
+                if self.to_f64()? > i32::MAX as f64 || self.to_f64()? < i32::MIN as f64 =>
+            {
                 Err(ODBCError::IntegralTruncation(x.to_string()))
             }
             _ => self.to_i64().map_or_else(
```
Comment (nbagnard) [inline odbc/src/api/data.rs:163]: umh, I am wondering if this is correct for string to i32 if the string represent a value outside of the i32 boundaries? Similarly to the case above `Bson::Int64(x) if *x > i32::MAX as i64 || *x < i32::MIN as i64`

**Review by nbagnard — CHANGES_REQUESTED** [review]
The conversion edge cases are looking good, except for one case that I think is supposed to fail.
I am also not sure why we are testing some conversions but not all.
The remaining comments are minor and more on code style/testing patterns we code use for easier maintenance later on.

File: odbc/src/api/data.rs [hunk 6]
```diff
@@ -108,14 +108,15 @@ impl IntoCData for Bson {
...
-            Bson::Double(f) => Ok((
+            Bson::Double(f) if *f < i64::MAX as f64 && *f > i64::MIN as f64 => Ok((
                 *f as i64,
                 if f.floor() != *f {
                     Some(ODBCError::FractionalTruncation(f.to_string()))
                 } else {
                     None
                 },
             )),
+            Bson::Double(f) => Err(ODBCError::IntegralTruncation(f.to_string())),
```
[the remark this answers is not in the document]
Comment (terakilobyte) [reply odbc/src/api/data.rs]: I've refactored these so only one branch is used in the guard, and bounds checking is performed within.

File: odbc/src/api/data.rs [hunk 7]
```diff
@@ -1254,15 +1261,60 @@ mod unit {
 
         #[test]
         fn positive_values_without_truncation_to_i32_succeed() {
-            let bson_double_to_i32 = Bson::Double(42.).to_u32();
-            let bson_int64_to_i32 = Bson::Int64(42).to_u32();
-            let bson_int32_to_i32 = Bson::Int32(42).to_u32();
+            let bson_double_to_i32 = Bson::Double(42.).to_i32();
```
[the remark this answers is not in the document]
Comment (terakilobyte) [reply odbc/src/api/data.rs]: I've refactored the tests so that the main supported types regarding bounds checking are handled - strings, int64, and int32. Notably, doubles are missing, but strings are converted to doubles and handled that way.

File: odbc/src/api/data.rs (hunk 5 above)
Comment (terakilobyte) [reply odbc/src/api/data.rs:170]: I don't like how decimals are checked and plan to revisit them when the Rust DBX team implements the type entirely to do proper bounds checking.

**Review by nbagnard — APPROVED** [review]
LGTM! 💯 
Nice refactoring, and elegant way to handle all the conversions, especially string data to C numeric 😊

File: odbc/src/api/data.rs [hunk 8]
```diff
@@ -1141,6 +1253,237 @@ mod unit {
...
+                    Ok(r) => unreachable!("should not have had an Ok {:?}", r),
+                }
+            };
+        }
+
+        macro_rules! test_it {
+            ($bson:expr,$v:expr) => {
+                $v.iter().for_each(
+                    |(method, expected, test, info)| match (method, test, info) {
+                        (&"i64", Ok(()), info) => {
```
Comment (pmeredit) [inline odbc/src/api/data.rs:1306]: I think the `&` here is redundant since "i64" is already a &str. Kind of surprised that clippy didn't complain, so maybe I'm wrong :D

**Review by pmeredit — APPROVED** [review]
LGTM

**Comment by terakilobyte** [comment]
evergreen merge
